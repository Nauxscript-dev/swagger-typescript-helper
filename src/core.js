export function getSchemaByRef(ref, schemaData) {
  if (!ref || typeof ref !== 'string')
    return undefined
  const keys = ref.split('/')
  let r = schemaData
  for (const key of keys) {
    if (r[key])
      r = r[key]
  }
  return r === schemaData ? undefined : r
}

export function generateFnHelper(type) {
  if (type === 'object')
    return generateObjectType

  if (type === 'array')
    return generateArrayType

  return generateBasicType
}

//
export function generateBasicType(propertyRaw) {
  const { type } = propertyRaw
  let typeName = type
  if (type === 'integer')
    typeName = 'number'

  return {
    typeName,
    deps: [],
  }
}

export function generateObjectType(propertyRaw, key, schemaData = apiDocsResponse) {
  const { $ref } = propertyRaw
  const schema = getSchemaByRef($ref, schemaData)

  if (!schema || !schema.properties)
    throw new Error('generateObjectType inner error')

  let typeValue = ''

  const deps = []

  for (const key in schema.properties) {
    if (Object.hasOwnProperty.call(schema.properties, key)) {
      const element = schema.properties[key]
      const elementType = element.type || 'object'
      const generateFn = generateFnHelper(elementType)
      const { typeName, statementContent, deps: innerDeps } = generateFn(element, key, schemaData)
      if (statementContent)
        deps.push(statementContent)

      if (innerDeps.length)
        deps.push(...innerDeps)

      typeValue += `\n${key}: ${typeName}`
    }
  }

  const typeName = `${capitalizeFirstLetter(key)}Type`

  typeValue = `{${typeValue}
  }`

  return {
    typeName,
    statementContent: `${generateHeadStr('object', typeName)}${typeValue}`,
    deps,
  }
}

// type === 'array'
export function generateArrayType(propertyRaw, key, schemaData = apiDocsResponse) {
  if (!propertyRaw.items) {
    if (!schema || !schema.properties)
      throw new Error('generateArrayType inner error')
  }

  const deps = []
  const type = propertyRaw.items.type || 'object'
  const generateFn = generateFnHelper(type)
  const { typeName, statementContent, deps: innerDeps } = generateFn(propertyRaw.items, `${key}Row`, schemaData)

  if (statementContent)
    deps.push(statementContent)

  if (innerDeps.length)
    deps.push(...innerDeps)

  return {
    typeName: `${typeName}[]`,
    deps,
  }
}

export function generateInterface(schema, name, wrapperType) {
  if (!schema)
    return 'any'

  // store the inner generated type
  const deps = []

  // gennerate head, such as:
  // "type xxx = "
  // "interface xxx "
  let headStr = generateSarter(wrapperType, name)

  const { type } = schema
  // object
  if (type === 'object') {
    const { typeValue } = generateObjectType(schema, name, deps)
    headStr += `${typeValue} \n`
  }

  // array
  if (type === 'array') {
    const { typeValue } = generateArrayType(schema, name, deps)
    headStr += `${typeValue} \n`
  }

  // eslint-disable-next-line no-console
  console.log(headStr, deps)
}

function capitalizeFirstLetter(string) {
  return string.charAt(0).toUpperCase() + string.slice(1)
}

function generateHeadStr(type, name) {
  if (type === 'object')
    return `\ninterface ${name} `

  if (type === 'array')
    return `\ntype ${name} = `

  if (type === 'enum')
    return `\ntype ${name} = `
}
