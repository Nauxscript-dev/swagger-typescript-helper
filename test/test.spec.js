import { test, expect } from 'vitest';
import { generateObjectType } from '../core';

test('simple object type', () => {
  const result = generateObjectType({
    $ref: '#/components/schemas/a.test.data'
  }, 'hello', {
    components: {
      schemas: {
        'a.test.data': {
          description: 'test schema',
          properties: {
            id: {
              description: "id string",
              type: "string"
            },
            name: {
              description: "name string",
              type: "string"
            }
          },
          required: ["id"],
          type: "object"
        }
      }
    }
  }) 
  expect(result).matchSnapshot()
})

test('nested simple object in an object type', () => {
  const result = generateObjectType({
    $ref: '#/components/schemas/a.test.data'
  }, 'hello', {
    components: {
      schemas: {
        'a.test.data': {
          description: 'test schema',
          properties: {
            id: {
              description: "id string",
              type: "string"
            },
            name: {
              $ref: '#/components/schemas/b.test.data' 
            }
          },
          type: "object"
        },
        'b.test.data': {
          description: 'test schema 2',
          properties: {
            type: {
              type: "string",
              description: "type string",
            },
          },
          type: "object" 

        }
      }
    }
  }) 
  expect(result).matchSnapshot()
})

test('nested simple array in an object type', () => {
  const result = generateObjectType({
    $ref: '#/components/schemas/a.test.data'
  }, 'hello', {
    components: {
      schemas: {
        'a.test.data': {
          description: 'test schema',
          properties: {
            id: {
              description: "id string",
              type: "string"
            },
            name: {
              description: 'object inner array',
              type: "array",
              items: {
                type: "string",
                description: 'array inner string'
              }
            }
          },
          type: "object"
        },
      }
    }
  }) 
  expect(result).matchSnapshot()
})

test('nested complex array in an object type', () => {
  const result = generateObjectType({
    $ref: '#/components/schemas/a.test.data'
  }, 'hello', {
    components: {
      schemas: {
        'a.test.data': {
          description: 'test schema',
          properties: {
            id: {
              description: "id string",
              type: "string"
            },
            name: {
              description: 'object inner array',
              type: "array",
              items: {
                $ref: '#/components/schemas/b.test.data' 
              }
            }
          },
          type: "object"
        },
        'b.test.data': {
          description: 'test schema 2',
          properties: {
            type: {
              type: "string",
              description: "type string",
            },
          },
          type: "object" 
        }
      }
    }
  }) 
  expect(result).matchSnapshot()
})