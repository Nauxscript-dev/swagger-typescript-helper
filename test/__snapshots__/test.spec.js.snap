// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`nested complex array in an object type 1`] = `
{
  "deps": [
    "
interface NameRowType {
type: string
  }",
  ],
  "statementContent": "
interface HelloType {
id: string
name: NameRowType[]
  }",
  "typeName": "HelloType",
}
`;

exports[`nested simple array in an object type 1`] = `
{
  "deps": [],
  "statementContent": "
interface HelloType {
id: string
name: string[]
  }",
  "typeName": "HelloType",
}
`;

exports[`nested simple object in an object type 1`] = `
{
  "deps": [
    "
interface NameType {
type: string
  }",
  ],
  "statementContent": "
interface HelloType {
id: string
name: NameType
  }",
  "typeName": "HelloType",
}
`;

exports[`simple object type 1`] = `
{
  "deps": [],
  "statementContent": "
interface HelloType {
id: string
name: string
  }",
  "typeName": "HelloType",
}
`;
